### DISCUSSION TITLE
Avalonia asking explicit DataType even when specified

### DISCUSSION THREAD

**Discussion by Mrcubix** [AvaloniaUI/Avalonia#13318] [Q&A]
Trying to build a shelf of node builders, allowing users to create a shape from a click of a button.
Tried with 11.0.4 & 11.0.5

Avalonia started prompting me with the following error on every uses of ItemsControl :
```
Cannot parse a compiled binding without an explicit x:DataType directive to give a starting data type for bindings.
```
That doesn't make sense to me as everything used to work fine until something happened and broke everything?

The current axaml header doesn't contain it, but it throw this error on every ItemsControl.

```axaml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:gn="clr-namespace:TouchGestures.UX.Controls.Nodes"
             xmlns:containers="clr-namespace:TouchGestures.UX.Controls.Containers"
             xmlns:vm="clr-namespace:TouchGestures.UX.ViewModels"
             xmlns:nodes="clr-namespace:TouchGestures.UX.ViewModels.Controls.Nodes"
             xmlns:enhancements="clr-namespace:TouchGestures.UX.Controls.Enhancements"
             xmlns:itfn="clr-namespace:TouchGestures.UX.Interfaces.Nodes"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             x:DataType="vm:MainViewModel"
             x:Class="TouchGestures.UX.Views.MainView">
```

The inherited interface for all builders is implemented as follow :

```cs
public interface INodeBuilder : IDisposable
{
    /// <summary>
    ///   The name of the shape built by this builder.
    /// </summary>
    public string Name { get; init; }

    /// <summary>
    ///   The icon of the shape built by this builder.
    /// </summary>
    public IImage Icon { get; init; }

    /// <summary>
    ///   Builds the node view model.
    /// </summary>
    /// <returns><see cref="NodeViewModel"/></returns>
    public NodeViewModel Build();
}
```
and the collection for `NodeBuilders` is declared like so : 
```cs
public ObservableCollection<INodeBuilder> NodeBuilders { get; } = new()
{
    new RectangleNodeBuilder(),
    new CircleNodeBuilder()
}
```

The code you'll see below worked once, but then never compiled again, with nothing changed except using `ColumnDefinition` instead of `RowDefinition`

Here is the seemingly faulty axaml code in question

```axaml
      <ItemsControl ItemsSource="{Binding NodeBuilders}">
        <ItemsControl.ItemsPanel>
          <ItemsPanelTemplate>
            <Grid />
          </ItemsPanelTemplate>
        </ItemsControl.ItemsPanel>
        <ItemsControl.ItemTemplate>
          <DataTemplate DataType="itfn:INodeBuilder">
            <Button Command="{Binding Build}" Background="#2A3531" >
              <Grid>
                <Grid.ColumnDefinitions>
                  <ColumnDefinition Width="Auto" SharedSizeGroup="Icons" />
                  <ColumnDefinition Width="*" />
                </Grid.ColumnDefinitions>

                <Image Source="{Binding Icon}" Grid.Column="0" />
                <TextBlock Text="{Binding Name}" Grid.Column="1" />
              </Grid>
            </Button>
          </DataTemplate>
        </ItemsControl.ItemTemplate>
      </ItemsControl>
```
where `itfn` is the defined namespace alias in the xml file.

As far as I'm aware, if it worked once, it should work again without any issues, but somehow, it doesn't.
Any ideas?

**Comment by maxkatz6** [comment]
Your code snipped should work as is, from what I see.
But even if you remove DataType="itfn:INodeBuilder", it should still work, as item type information should be available from ItemsSource="{Binding NodeBuilders}" (since NodeBuilders has specific types and not just objects).

There could be some corner case that breaks the compiler, but we need a minimal repro to fix and test it.

**Reply by Mrcubix** [reply]
Tried making a minimal reproduction using just that part alone but for some reasons, it doesn't occur there, and it doesn't seem to be a caching issue since removing obj and bin in all avalonia dependent project doesn't fix it either.
